### PR TITLE
Add new built-in type, dict

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
For now dicts are immutable, and do not expose any methods,
they however support "native" syntax such as [] lookup,
and foreach iterating, and can be printed.

If it is desirable, the methods implemented by https://github.com/mesonbuild/meson/pull/1295 could be added later.

Here are the new syntax elements:

Creating a dictionary:

```
dict = {'foo': 'bar', 'baz': 42}
```

Looking up an item:

```
foo = dict['foo']
```

Iterating:

```
foreach key, value: dict
  # Do something
endforeach
```

### CODE REVIEW

File: test cases/common/198 dict/meson.build [hunk 1]
```diff
@@ -0,0 +1,12 @@
+project('dict test', 'c')
+
+dict = {foo : 'bar',
```
Comment (sarum9in) [inline test cases/common/198 dict/meson.build]: How do I specify key with spaces?

File: test cases/common/198 dict/meson.build [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+        baz : 'foo',
+        foobar: 'baz'}
```
Comment (sarum9in) [inline test cases/common/198 dict/meson.build]: Also it might make sense to have assignment operator similar to `+=` for arrays. It would create a copy of the dict with additional/overridden value.

File: test cases/common/198 dict/meson.build (hunk 1 above)
Comment (MathieuDuponchelle) [reply test cases/common/198 dict/meson.build]: you don't for now, it was a bit late and I just reused the kwargs code path, I think I'll change that :)

File: test cases/common/198 dict/meson.build (hunk 2 above)
Comment (MathieuDuponchelle) [reply test cases/common/198 dict/meson.build]: this hinges on the decision of whether a dict should be a mutable object or not, I didn't want to make arguing my case more complicated than strictly needed tbh ;)

**Comment by MathieuDuponchelle** [comment]
my use case is pairing component names and lists of sources, here's a before / after comparison:

https://pastebin.com/raw/iEtNY5Bw and https://pastebin.com/raw/8iys5zuL

File: test cases/common/198 dict/meson.build (hunk 1 above)
Comment (MathieuDuponchelle) [reply test cases/common/198 dict/meson.build]: Fixed

Comment (MathieuDuponchelle) [reply test cases/common/198 dict/meson.build]: @nirbheek , you don't for now, I don't know if we'll want that later, I've just made it so you can only use literal strings as keys

Comment (nirbheek) [reply test cases/common/198 dict/meson.build]: Probably better to match Python's style and have the syntax be `{'foo': 'bar'}` then? So we can extend it later.

Comment (MathieuDuponchelle) [reply test cases/common/198 dict/meson.build]: yes that's what I did, check the updated tests ;)

**Comment by haasn** [comment]
I've long wished for a similar thing to this, with related syntax. It may be possible we could combine both of our use cases. Specifically, what I'm interested in is the ability to define (immutable) objects using a syntax like this:

```
foo = {
  name: 'foobar',
  whatever: True,
  depends_on: [ 'batbaz' ]
  files: [ 'foo.c', 'bar.c' ]
}

# can be accessed as foo.name or foo.files
```

The only major change between yours and mine is that I find it redundant to also need to escape the key name as a string, when to me it seems like an identifier would make more sense (not a string). Also, indexing it could just be using subscript notation (similar to how e.g. `cc.find_library` works), rather than needing to use the clunky array indexing (imagine writing `cc['find_library']`!).

It seems like the only thing you really need is the ability to iterate through these object-types in a key/value manner. That should make us both happy, no?

**Comment by haasn** [comment]
> subscript notation

And by that I mean dot notation. As for a justification, consider how else one would write my example. It would have to be as an ugly amalgation of arrays, e.g.:

```
foo = [
  'foobar', # name
  True, # whatever
  [ 'batbaz' ] # depends_on,
  [ 'foo.c', 'bar.c' ] # files
]

# can be accessed as foo[0] or foo[3]
```

The differences here are obvious and striking:
1. Using named fields immediately documents which field is which, without needing to introduce extra comments for this purpose
2. You can leave away fields that you don't specifically need for an object, so if you have 10+ supported fields you don't need to include 10 extra dummy “null” options for the fields you're not interested in for that component.
3. There's no risk of potentially confusing the array indices with each other. I know when I write `foo.files` that I'm getting the list of files, and not accidentally the `depends_on` field because I was off by one in my memory of which ID is which.
4. You can easily add new optional fields later on without having to make sure to adjust every single other component to make sure the array indices don't break.

In general it's just a whole lot nicer and easier to work with. Programming languages have these sorts of record constructs for a reason, there's no need to justify it further than this example IMO. Otherwise you might as well argue that the only data type we need is strings and we'd be back to writing shell scripts. (Ironically, even shell languages like bash or zsh support key/value dictionaries these days)

**Comment by haasn** [comment]
I will also re-iterate here that I think a valid alternative for my use case would be the ability to define your own `constructor` functions, like perhaps:

```
object Component(name, whatever: False, depends_on: [], files: [])

foo = Component('foo'
    whatever: True,
    depends_on: [ 'batbaz' ]
)

# foo.name => 'foo'
# foo.files => defaults to []
```

**Comment by MathieuDuponchelle** [comment]
@haasn, ugly amalgamations of arrays is indeed one of the things a dict object would help us avoid, I have encountered this pattern quite a few times in gstreamer's build definitions where we check array.length() and determine the nature of "optional" components based on their positions.

Re ids vs strings as keys, my first draft indeed used ids, but as was pointed out that seemed like an unneeded restriction, which I removed.

Re objects, I think that's really a different topic, I propose we stick to discussing dicts here :)

File: test cases/common/198 dict/meson.build (hunk 2 above)
Comment (sarum9in) [reply test cases/common/198 dict/meson.build]: Sure, we could introduce it later if needed, let's keep it simple for now. It would not make dicts mutable though as semantics would be same as in arrays.

**Comment by MathieuDuponchelle** [comment]
Here is a typical example of the "ugly accumulation of arrays" case:

https://github.com/GStreamer/gst-plugins-base/blob/master/tests/check/meson.build#L4

then:

https://github.com/GStreamer/gst-plugins-base/blob/master/tests/check/meson.build#L133

in other cases, there are even more "optional" items, which end up being annoying to write and difficult to understand.

With a dict, one could instead have:

```
['foo'],
['bar', {'skip': false}]
['baz', {deps: [dep]}]
```

This is pretty much the declarative equivalent to keyword args on function calls.

**Comment by tp-m** [comment]
I completely agree with you Jussi that it needs a good use case and shouldn't be added willy-nilly just because it seems nice to have in general.

However, there is a point at which certain arrays become too unwieldy and simply too ugly. This hinders readability and maintainability of the Meson build definitions.

It's also needed for performance reasons in our ffmpeg meson build which has long lists of various types of components that needs to be cross-checked against other components/things. This means we have to iterate arrays with hundreds of items and for each entry lookup others in the array, which means iterating through these arrays until each entry is found. This is suboptimal. We could abuse set_variable() and get_variable(), but that feels wrong and hacky, and I'd rather improve Meson than hack around it.

Most of all a dict is simply the right tool for the job in this case, and in many other cases. I have a number of projects where I would use it.

It also seems perfectly in line with all aspects of the Meson philosophy as I understand it. It's declarative. It's intuitive. It can be iterated just like arrays. It's fairly light on the implementation side.

I think we should only support immutable dicts for now, however.

Would be great if we could get this in.

**Comment by MathieuDuponchelle** [comment]
For what it's worth, once a dict object is available, using it comes naturally, see for example that port I made yesterday of pixman: https://github.com/centricular/pixman/blob/meson/meson.build

**Comment by MathieuDuponchelle** [comment]
ping @jpakkane :)

**Comment by jpakkane** [comment]
Overall this seems like a real world problem that is hard to do with arrays of arrays so we should definitely do something to make it better. This is a fairly good approach to this issue (but if we come up with something else that works we could use that as well).

The interesting thing here is the usage portion of the libavcodec example. Can you post a comparison on how that works? That is, how do you then go over all the choices and pick the ones that match whatever criteria you may have?

> in our ffmpeg meson build which has long lists of various types of components that needs to be cross-checked against other components/things. This means we have to iterate arrays with hundreds of items and for each entry lookup others in the array, which means iterating through these arrays until each entry is found.

Could you post a simplified ELI5 version of the thing that you need to do here? Specifically what is being iterated over and how?

**Comment by xclaesse** [comment]
Another potential use-case: I think my PR #3320 could be rewritted using dict.

File: mesonbuild/mparser.py [hunk 3]
```diff
@@ -664,10 +706,17 @@ def foreachblock(self):
         t = self.current
         self.expect('id')
         varname = t
+        varnames = [t]
+
+        if (self.accept('comma')):
```
Comment (jpakkane) [inline mesonbuild/mparser.py]: Unnecessary parentheses here.

**Comment by jpakkane** [comment]
The dict should probably have a method `.get(varname, default_value)` that returns `default_value` if `varname` is not a key that exists in the dictionary.

Also needs some more thorough tests, such as:

 - `{}` creates an empty dictionary
 - `{3, 'foo': 'bar'}` should fail
 - as should `{'foo': 'bar', 3}

**Comment by MathieuDuponchelle** [comment]
@jpakkane , I was waiting for an initial approval before fleshing this out, I do have a commit on top of that that implements get and contains() (should it be named `has()` ?), which I adapted from https://github.com/mesonbuild/meson/pull/1295 .

If you reckon we can go ahead with that (woot :)), I will address your comments and also update the documentation :)

File: mesonbuild/mparser.py (hunk 3 above)
Comment (MathieuDuponchelle) [reply mesonbuild/mparser.py]: I blame C

**Comment by jpakkane** [comment]
>  contains() (should it be named has()

Maybe `has_key` for explicitness?

The unit test that loops over the dict should also calculate the amount of iterations it does and assert that it is 3.

> If you reckon we can go ahead with that (woot :)), I will address your comments and also update the documentation 

Go for it.

**Comment by MathieuDuponchelle** [comment]
Went for it :)

**Comment by ignatenkobrain** [comment]
Flake8 detected 6 issues on 5d56b4353e778875528c582626bcec9b0f245b8a
Visit https://sideci.com/gh/19784232/pull_requests/3490 to review the issues.

File: run_unittests.py [hunk 4]
```diff
@@ -582,6 +582,7 @@ def init(self, srcdir, extra_args=None, default_args=True, inprocess=False):
         if inprocess:
             try:
                 (returncode, out, err) = run_configure(self.meson_mainfile, self.meson_args + args + extra_args)
+                print (out)
```
Comment (jpakkane) [inline run_unittests.py]: Stray?

**Comment by jpakkane** [comment]
Still missing the failing tests, especially the duplicate key one.

Should perhaps also add a failing test for someone trying to build a dict with integer keys, such as `x = {3: 'c'}`.

**Comment by jpakkane** [comment]
> It's in unit tests, can't be tested from a project file

Well you can, that is what `failing` tests are for. But on the other hand that one is possibly even better and I did not even realize that functionality had been added in `run_unittests`. I'm glad someone knows about these things at least. :)

**Comment by QuLogic** [comment]
This would fix #2463.

**Comment by jpakkane** [comment]
Looking pretty good. The documentation needs to be amended to contain *available since 0.47.0* in all the places dicts are discussed but other than that this seems to be ready.

**Comment by jpakkane** [comment]
The time has come. Thanks to everyone who participated in this.
